controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/AbstractRequestProcessor.java

```diff
@@ -0,0 +1,184 @@
+        CompletableFuture<Void> resultFuture = new CompletableFuture<>();
+
+        OperationContext context = streamMetadataStore.createContext(scope, stream);
+        suppressException(streamMetadataStore.getWaitingRequestProcessor(scope, stream, context, executor), null,
+                "Exception while trying to fetch waiting request. Logged and ignored.")
+                .thenAccept(waitingRequestProcessor -> {
+                    if (waitingRequestProcessor == null || waitingRequestProcessor.equals(getProcessorName())) {
+                        withRetries(() -> task.execute(event), executor)
+                                .whenComplete((r, ex) -> {
+                                    if (ex != null && writeBackPredicate.test(ex)) {
+                                        suppressException(streamMetadataStore.createWaitingRequestIfAbsent(scope, stream, getProcessorName(), context, executor),
+                                                null, "Exception while trying to create waiting request. Logged and ignored.")
+                                                .thenCompose(ignore ->  retryIndefinitelyThenComplete(() -> task.writeBack(event), resultFuture, ex));
+                                    } else {
+                                        // Processing was done for this event, whether it succeeded or failed, we should remove
+                                        // the waiting request if it matches the current processor.
+                                        // If we dont delete it then some other processor will never be able to do the work.
+                                        // So we need to retry indefinitely until deleted.
+                                        retryIndefinitelyThenComplete(() -> streamMetadataStore.deleteWaitingRequestConditionally(scope,
+                                                stream, getProcessorName(), context, executor), resultFuture, ex);
+                                    }
+                                });
+                    } else {
+                        log.debug("Found another processing requested by a different processor {}. Will postpone the event {}.", waitingRequestProcessor, event);
+                        // This is done to guarantee fairness. If another processor has requested for processing
+                        // on this stream, we will back off and postpone the work for later.
+                        retryIndefinitelyThenComplete(() -> task.writeBack(event), resultFuture,
+                                StoreException.create(StoreException.Type.OPERATION_NOT_ALLOWED, "Postponed "
+                                        + event + " so that waiting processor" + waitingRequestProcessor + " can work. "));
+                    }
+                });
+
+        return resultFuture;
+    }
+
+    private <R> CompletableFuture<R> suppressException(CompletableFuture<R> future, R returnOnException, String message) {
```

Comment:
You can use `Futures.exceptionallyExpecting` which does the same thing ... minus the logging.
You can sneak in the logging in the Predicate that you pass into that method.

Reply:
The level of indirection that this method is introducing seems unnecessary. Could we simply call this directly:

```
Futures.exceptionallyExpecting(future, e -> {
            log.debug(message, e);
            return true;
        }, returnOnException);
```

Reply:
We need to suppress exception in three places in the code. A common private method seemed cleaner and makes the code more readable.